lib/Relation.php

```diff
@@ -0,0 +1,557 @@
+<?php
+/**
+ * Interface
+ *
+ * @package ActiveRecord
+ */
+
+namespace ActiveRecord;
+
+use ActiveRecord\Exception\RecordNotFound;
+use ActiveRecord\Exception\ValidationsArgumentError;
+
+/**
+ * @template TModel of Model
+ *
+ * @phpstan-import-type RelationOptions from Types
+ */
+class Relation
+{
+    /**
+     * @var array<string>
+     */
+    private array $alias_attribute;
+
+    /**
+     * @var class-string
+     */
+    private string $className;
+
+    private ?Table $tableImpl = null;
+
+    /**
+     * @var RelationOptions
+     */
+    private array $options = [];
+
+    /**
+     * @param class-string    $className
+     * @param array<string>   $alias_attribute
+     * @param RelationOptions $options
+     */
+    public function __construct(string $className, array $alias_attribute, array $options = [])
+    {
+        $this->options = $options;
+        $this->alias_attribute = $alias_attribute;
+        $this->className = $className;
+    }
+
+    /**
+     * @return Table<TModel>
+     */
+    private function table(): Table
+    {
+        if (null === $this->tableImpl) {
+            $this->tableImpl = Table::load($this->className);
+        }
+
+        return $this->tableImpl;
+    }
+
+    /**
+     * @return Relation<TModel>
+     */
+    public function select(string $columns): Relation
+    {
+        $this->options['select'] = $columns;
+
+        return $this;
+    }
+
+    /**
+     * Performs JOINs on +args+. The given symbol(s) should match the name of
+     * the association(s).
+     *
+     * User::joins('posts') // SELECT "users".*
+     *                      // FROM "users"
+     *                      // INNER JOIN "posts" ON "posts"."user_id" = "users"."id"
+     *
+     * // Multiple joins:
+     *
+     * User::joins(['posts', 'account']) // SELECT "users".*
+     *                                   // FROM "users"
+     *                                   // INNER JOIN "posts" ON "posts"."user_id" = "users"."id"
+     *                                   // INNER JOIN "accounts" ON "accounts"."id" = "users"."account_id"
+     *
+     * @param string|array<string> $joins
+     *
+     * @returns Relation<TModel>
+     */
+    public function joins(string|array $joins): Relation
+    {
+        $this->options['joins'] = $joins;
+
+        return $this;
+    }
+
+    /**
+     * @return Relation<TModel>
+     */
+    public function order(string $order): Relation
+    {
+        $this->options['order'] = $order;
+
+        return $this;
+    }
+
+    /**
+     * Specifies a limit for the number of records to retrieve.
+     *
+     * User::limit(10) // generated SQL has 'LIMIT 10'
+     * User::limit(10)->limit(20) # generated SQL has 'LIMIT 20'
+     *
+     * @return Relation<TModel>
+     */
+    public function limit(int $limit): Relation
+    {
+        $this->options['limit'] = $limit;
```

Comment:
Bug in $this->take() case where it calls limit(null): should be:

if ($limit !== null) {
   $this->options['limit'] = $limit;
}

Reply:
Mm, not sure about that one; `null` is not currently one of the legal values on this method, so PHP will throw before we ever get as far as writing it.